File: tom_som_python_runtime/tom_som_runtime/spec_meta_bridge.py
"""Bridge from the meta-JSON class graph (:class:`SpecModel`) to the canonical
metadata tree (:class:`SomMetaTree`, DR1 §3.1) — a faithful port of
`tom_som_dart_runtime/lib/src/spec_meta_bridge.dart`.

The generated facades (DR8) emit populated :class:`SomMetaTree` s directly;
every consumer that only has the exported spec-model meta-data (the
conformance harness, tooling) builds its tree here. The expansion follows the
same walk :meth:`SpecReflection.resolve` performs — crucially, a node's
:attr:`SomMetaNode.section_id` is the **field-level** id exactly as exported
(``field.section_id``), so the tree's path grammar stays byte-compatible with
the paths a :class:`SpecDocument` is keyed by. (DR1 §2.2's "field id, else
target-class id" resolution is applied at *export* time by the model exporter
/ DR8 generator, not re-derived here.)

Recursive class references become terminal re-entry nodes
(:attr:`SomMetaNode.recursive`), mirroring how the generated chains terminate.
"""

from __future__ import annotations

from typing import Optional

from .spec_meta import (
    SomContentTypeMeta,
    SomDocMeta,
    SomFormFieldMeta,
    SomFormMeta,
    SomMetaExtra,
    SomMetaKind,
    SomMetaNode,
    SomMetaTree,
    SomSecondLevelId,
)
from .spec_model import (
    SpecAnnotation,
    SpecClass,
    SpecField,
    SpecFieldKind,
    SpecModel,
)

#: Annotation names that have a dedicated :class:`SomMetaNode` slot; everything
#: else is captured losslessly into :attr:`SomMetaNode.extra`.
_SLOTTED_ANNOTATIONS = {
    "SectionId",
    "SectionIdPattern",
    "SerializationOrder",
    "Min",
    "Unused",
    "ContentType",
    "ContentHelp",
    "Headline",
    "Comment",
    "Form",
    "Document",
    "MapsTo",
    "DetailedIn",
    "SecondLevelSectionId",
}


def build_som_meta_tree(
    model: SpecModel, root_type: Optional[str] = None
) -> SomMetaTree:
    """Builds the wired :class:`SomMetaTree` for one document root of *model*.

    When *root_type* is omitted the model's first root is used; otherwise the
    root is looked up via :meth:`SpecModel.root_by_type` (raising
    :class:`ValueError` for an unknown type). Children are ordered by
    ``@SerializationOrder`` (annotated members first, then declaration order),
    which is the sibling emission order of the hierarchical YAML format (DR1
    §2.3)."""
    if root_type is None and model.roots:
        root = model.roots[0]
    else:
        root = model.root_by_type(root_type or "")
    cls = model.class_named(root.type)
    root_node = SomMetaNode(
        class_name=root.type,
        section_id=root.section_id or (cls.section_id if cls else None),
        class_section_id=cls.section_id if cls else None,
        kind=SomMetaKind.SECTION,
        type_name=root.type,
        headline=cls.headline if cls else None,
        doc_comment=root.doc or (cls.doc if cls else None),
        class_doc_comment=cls.doc if cls else None,
        maps_to=cls.maps_to if cls else None,
        detailed_in=cls.detailed_in if cls else None,
        document=SomDocMeta(
            name=root.title,
            description=root.description or "",
            based_on=_based_on(cls),
        ),
        second_level_ids=_second_level_ids(cls.annotations if cls else []),
        extra=_extras(cls.annotations if cls else []),
        children=[] if cls is None else _child_nodes(model, cls, {root.type}),
    )
    return SomMetaTree(root_node)


def _based_on(cls: Optional[SpecClass]) -> list[str]:
    ann = cls.annotation("Document") if cls else None
    raw = ann.argument("basedOn") if ann else None
    if isinstance(raw, list):
        return [str(e) for e in raw]
    return []


def _child_nodes(
    model: SpecModel, cls: SpecClass, stack: set[str]
) -> list[SomMetaNode]:
    fallback = 1 << 30
    indexed = sorted(
        enumerate(cls.fields),
        key=lambda e: (
            e[1].serialization_order
            if e[1].serialization_order is not None
            else fallback,
            e[0],
        ),
    )
    return [_field_node(model, cls, f, stack) for _, f in indexed]


def _field_node(
    model: SpecModel, owner: SpecClass, field: SpecField, stack: set[str]
) -> SomMetaNode:
    kind = _KIND_OF[field.kind]
    target: Optional[SpecClass] = None
    recursive = False
    children: list[SomMetaNode] = []
    element_node: Optional[SomMetaNode] = None

    if field.kind in (SpecFieldKind.COMPLEX, SpecFieldKind.SECTION):
        target = model.class_named(field.type)
        if target is not None:
            if target.name in stack:
                recursive = True
            else:
                stack.add(target.name)
                children = _child_nodes(model, target, stack)
                stack.remove(target.name)
    elif field.kind is SpecFieldKind.LIST and field.element_is_complex:
        element = model.class_named(field.element_type)
        if element is not None:
            element_children: list[SomMetaNode] = []
            element_recursive = False
            if element.name in stack:
                element_recursive = True
            else:
                stack.add(element.name)
                element_children = _child_nodes(model, element, stack)
                stack.remove(element.name)
            element_node = SomMetaNode(
                class_name=element.name,
                class_section_id=element.section_id,
                kind=SomMetaKind.COMPLEX,
                type_name=element.name,
                headline=element.headline,
                doc_comment=element.doc,
                class_doc_comment=element.doc,
                maps_to=element.maps_to,
                detailed_in=element.detailed_in,
                recursive=element_recursive,
                children=element_children,
            )

    content_type = None
    if field.content_type is not None:
        ct_ann = field.annotation("ContentType")
        content_type = SomContentTypeMeta(
            type=field.content_type,
            description=str(
                (ct_ann.argument("description") if ct_ann else None) or ""
            ),
        )

    comment_ann = field.annotation("Comment")
    comment_text = str(
        (comment_ann.argument("text") if comment_ann else None) or ""
    )

    form = None
    if field.kind is SpecFieldKind.FORM:
        form = SomFormMeta(
            fields=[
                SomFormFieldMeta(
                    name=ff.name,
                    type_name=ff.type,
                    description=ff.label,
                    required=ff.required,
                    hint=ff.hint,
                    role=ff.role,
                    initial=ff.initial,
                    order=i,
                )
                for i, ff in enumerate(field.form_fields)
            ]
        )

    return SomMetaNode(
        class_name=target.name if target else owner.name,
        member_name=field.name,
        section_id=field.section_id,
        class_section_id=target.section_id if target else None,
        section_id_pattern=field.section_id_pattern,
        kind=kind,
        type_name=field.type or field.element_type or field.enum_type
        or "String",
        serialization_order=field.serialization_order,
        min=field.min,
        unused=field.annotation("Unused") is not None,
        content_type=content_type,
        content_help=field.help,
        headline=field.headline
        if field.headline is not None
        else (target.headline if target else None),
        comment=comment_text if comment_text else None,
        doc_comment=field.doc or (target.doc if target else None),
        class_doc_comment=target.doc if target else None,
        form=form,
        maps_to=target.maps_to if target else None,
        detailed_in=target.detailed_in if target else None,
        second_level_ids=_second_level_ids(field.annotations),
        extra=_extras(field.annotations),
        recursive=recursive,
        children=children,
        element_node=element_node,
    )


_KIND_OF = {
    SpecFieldKind.LIST: SomMetaKind.LIST,
    SpecFieldKind.FORM: SomMetaKind.FORM,
    SpecFieldKind.SECTION: SomMetaKind.SECTION,
    SpecFieldKind.CONTENT: SomMetaKind.CONTENT,
    SpecFieldKind.ENUM: SomMetaKind.ENUM_VALUE,
    SpecFieldKind.COMPLEX: SomMetaKind.COMPLEX,
    SpecFieldKind.SCALAR: SomMetaKind.SCALAR,
}


def _second_level_ids(
    annotations: list[SpecAnnotation],
) -> list[SomSecondLevelId]:
    return [
        SomSecondLevelId(
            document_class=str(a.argument("documentClass") or ""),
            id=str(a.argument("id") or ""),
        )
        for a in annotations
        if a.name == "SecondLevelSectionId"
    ]


def _extras(annotations: list[SpecAnnotation]) -> list[SomMetaExtra]:
    return [
        SomMetaExtra(annotation=a.name, args=a.arguments)
        for a in annotations
        if a.name not in _SLOTTED_ANNOTATIONS
    ]

File: tom_som_typescript_runtime/src/spec_meta_bridge.ts
/**
 * Bridge from the meta-JSON class graph ({@link SpecModel}) to the canonical
 * metadata tree ({@link SomMetaTree}, DR1 §3.1) — a faithful port of
 * `tom_som_dart_runtime/lib/src/spec_meta_bridge.dart` (and the JavaScript
 * `spec_meta_bridge.js`).
 *
 * The generated facades (DR8) emit populated {@link SomMetaTree}s directly;
 * every consumer that only has the exported spec-model meta-data (the
 * conformance harness, tooling) builds its tree here. The expansion follows
 * the same walk {@link SpecReflection.resolve} performs — crucially, a node's
 * {@link SomMetaNode.sectionId} is the **field-level** id exactly as exported
 * (`field.sectionId`), so the tree's path grammar stays byte-compatible with
 * the paths a {@link SpecDocument} is keyed by. (DR1 §2.2's "field id, else
 * target-class id" resolution is applied at *export* time by the model
 * exporter / DR8 generator, not re-derived here.)
 *
 * Recursive class references become terminal re-entry nodes
 * ({@link SomMetaNode.recursive}), mirroring how the generated chains
 * terminate.
 */

import {
  SomContentTypeMeta,
  SomDocMeta,
  SomFormFieldMeta,
  SomFormMeta,
  SomMetaExtra,
  SomMetaKind,
  SomMetaKindValue,
  SomMetaNode,
  SomMetaTree,
  SomSecondLevelId,
} from './spec_meta';
import {
  SpecAnnotation,
  SpecClass,
  SpecField,
  SpecFieldKind,
  SpecFieldKindValue,
  SpecModel,
} from './spec_model';

/**
 * Annotation names that have a dedicated {@link SomMetaNode} slot; everything
 * else is captured losslessly into {@link SomMetaNode.extra}.
 */
const _SLOTTED_ANNOTATIONS = new Set([
  'SectionId',
  'SectionIdPattern',
  'SerializationOrder',
  'Min',
  'Unused',
  'ContentType',
  'ContentHelp',
  'Headline',
  'Comment',
  'Form',
  'Document',
  'MapsTo',
  'DetailedIn',
  'SecondLevelSectionId',
]);

const _KIND_OF: Record<SpecFieldKindValue, SomMetaKindValue> = {
  [SpecFieldKind.LIST]: SomMetaKind.LIST,
  [SpecFieldKind.FORM]: SomMetaKind.FORM,
  [SpecFieldKind.SECTION]: SomMetaKind.SECTION,
  [SpecFieldKind.CONTENT]: SomMetaKind.CONTENT,
  [SpecFieldKind.ENUM]: SomMetaKind.ENUM_VALUE,
  [SpecFieldKind.COMPLEX]: SomMetaKind.COMPLEX,
  [SpecFieldKind.SCALAR]: SomMetaKind.SCALAR,
};

/**
 * Builds the wired {@link SomMetaTree} for one document root of `model`.
 *
 * When `rootType` is omitted the model's first root is used; otherwise the
 * root is looked up via {@link SpecModel.rootByType} (throwing for an unknown
 * type). Children are ordered by `@SerializationOrder` (annotated members
 * first, then declaration order), which is the sibling emission order of the
 * hierarchical YAML format (DR1 §2.3).
 */
export function buildSomMetaTree(
  model: SpecModel,
  rootType: string | null = null,
): SomMetaTree {
  const root =
    rootType === null || rootType === undefined
      ? model.roots.length > 0
        ? model.roots[0]
        : model.rootByType('')
      : model.rootByType(rootType);
  const cls = model.classNamed(root.type);
  const rootNode = new SomMetaNode({
    className: root.type,
    sectionId:
      root.sectionId != null ? root.sectionId : cls !== null ? cls.sectionId : null,
    classSectionId: cls !== null ? cls.sectionId : null,
    kind: SomMetaKind.SECTION,
    typeName: root.type,
    headline: cls !== null ? cls.headline : null,
    docComment: root.doc != null ? root.doc : cls !== null ? cls.doc : null,
    classDocComment: cls !== null ? cls.doc : null,
    mapsTo: cls !== null ? cls.mapsTo : null,
    detailedIn: cls !== null ? cls.detailedIn : null,
    document: new SomDocMeta({
      name: root.title,
      description: root.description != null ? root.description : '',
      basedOn: _basedOn(cls),
    }),
    secondLevelIds: _secondLevelIds(cls !== null ? cls.annotations : []),
    extra: _extras(cls !== null ? cls.annotations : []),
    children: cls === null ? [] : _childNodes(model, cls, new Set([root.type])),
  });
  return new SomMetaTree(rootNode);
}

function _basedOn(cls: SpecClass | null): string[] {
  const ann = cls !== null ? cls.annotation('Document') : null;
  const raw = ann !== null ? ann.argument('basedOn') : null;
  if (Array.isArray(raw)) {
    return raw.map((e) => String(e));
  }
  return [];
}

function _childNodes(
  model: SpecModel,
  cls: SpecClass,
  stack: Set<string>,
): SomMetaNode[] {
  const fallback = 1 << 30;
  const indexed: Array<[number, SpecField]> = cls.fields.map((f, i) => [i, f]);
  indexed.sort((a, b) => {
    const oa = a[1].serializationOrder !== null ? a[1].serializationOrder : fallback;
    const ob = b[1].serializationOrder !== null ? b[1].serializationOrder : fallback;
    if (oa !== ob) {
      return oa - ob;
    }
    return a[0] - b[0];
  });
  return indexed.map(([, f]) => _fieldNode(model, cls, f, stack));
}

function _fieldNode(
  model: SpecModel,
  owner: SpecClass,
  field: SpecField,
  stack: Set<string>,
): SomMetaNode {
  const kind = _KIND_OF[field.kind];
  let target: SpecClass | null = null;
  let recursive = false;
  let children: SomMetaNode[] = [];
  let elementNode: SomMetaNode | null = null;

  if (field.kind === SpecFieldKind.COMPLEX || field.kind === SpecFieldKind.SECTION) {
    target = model.classNamed(field.type);
    if (target !== null) {
      if (stack.has(target.name)) {
        recursive = true;
      } else {
        stack.add(target.name);
        children = _childNodes(model, target, stack);
        stack.delete(target.name);
      }
    }
  } else if (field.kind === SpecFieldKind.LIST && field.elementIsComplex) {
    const element = model.classNamed(field.elementType);
    if (element !== null) {
      let elementChildren: SomMetaNode[] = [];
      let elementRecursive = false;
      if (stack.has(element.name)) {
        elementRecursive = true;
      } else {
        stack.add(element.name);
        elementChildren = _childNodes(model, element, stack);
        stack.delete(element.name);
      }
      elementNode = new SomMetaNode({
        className: element.name,
        classSectionId: element.sectionId,
        kind: SomMetaKind.COMPLEX,
        typeName: element.name,
        headline: element.headline,
        docComment: element.doc,
        classDocComment: element.doc,
        mapsTo: element.mapsTo,
        detailedIn: element.detailedIn,
        recursive: elementRecursive,
        children: elementChildren,
      });
    }
  }

  let contentType: SomContentTypeMeta | null = null;
  if (field.contentType !== null && field.contentType !== undefined) {
    const ctAnn = field.annotation('ContentType');
    contentType = new SomContentTypeMeta(
      field.contentType,
      String((ctAnn !== null ? ctAnn.argument('description') : null) || ''),
    );
  }

  const commentAnn = field.annotation('Comment');
  const commentText = String(
    (commentAnn !== null ? commentAnn.argument('text') : null) || '',
  );

  let form: SomFormMeta | null = null;
  if (field.kind === SpecFieldKind.FORM) {
    form = new SomFormMeta(
      field.formFields.map(
        (ff, i) =>
          new SomFormFieldMeta({
            name: ff.name,
            typeName: ff.type,
            description: ff.label,
            required: ff.required,
            hint: ff.hint,
            order: i,
          }),
      ),
    );
  }

  return new SomMetaNode({
    className: target !== null ? target.name : owner.name,
    memberName: field.name,
    sectionId: field.sectionId,
    classSectionId: target !== null ? target.sectionId : null,
    sectionIdPattern: field.sectionIdPattern,
    kind,
    typeName: field.type || field.elementType || field.enumType || 'String',
    serializationOrder: field.serializationOrder,
    min: field.min,
    unused: field.annotation('Unused') !== null,
    contentType,
    contentHelp: field.help,
    headline:
      field.headline != null
        ? field.headline
        : target !== null
          ? target.headline
          : null,
    comment: commentText !== '' ? commentText : null,
    docComment: field.doc || (target !== null ? target.doc : null),
    classDocComment: target !== null ? target.doc : null,
    form,
    mapsTo: target !== null ? target.mapsTo : null,
    detailedIn: target !== null ? target.detailedIn : null,
    secondLevelIds: _secondLevelIds(field.annotations),
    extra: _extras(field.annotations),
    recursive,
    children,
    elementNode,
  });
}

function _secondLevelIds(annotations: SpecAnnotation[]): SomSecondLevelId[] {
  return annotations
    .filter((a) => a.name === 'SecondLevelSectionId')
    .map(
      (a) =>
        new SomSecondLevelId(
          String(a.argument('documentClass') || ''),
          String(a.argument('id') || ''),
        ),
    );
}

function _extras(annotations: SpecAnnotation[]): SomMetaExtra[] {
  return annotations
    .filter((a) => !_SLOTTED_ANNOTATIONS.has(a.name))
    .map((a) => new SomMetaExtra(a.name, a.arguments));
}
